docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,247 @@
+
+Another shorter-term risk is that although ASO v2 is [officially supported by Microsoft](https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-support-model), it has not yet published a GA release. Until then, ASO may make breaking changes between pre-releases. The remaining work to be completed before ASO v2's first GA release is being tracked here: https://github.com/Azure/azure-service-operator/milestone/20.
+
+Further down the road, there is an apparent increased risk that CAPZ's lifespan may extend beyond that of its Azure interface when ASO is used instead of the API or SDK directly, as ASO has not yet been proven as a mission-critical interface to Azure to the same extent as Azure's REST API and Go SDK.
+
+Since ASO can be freely installed by users, there is also a risk that a CAPZ-installed ASO may conflict with a user-installed instance. Before CAPZ installs ASO, it should verify that ASO is already either not installed or installed and configured as expected by CAPZ.
+
+The difference between CAPZ's and ASO's metrics stories may also be considered a risk. CAPZ's metrics currently include traces that measure how long each step in a reconciliation takes, including all Azure API calls. Once those equivalent API calls are being done from ASO instead of CAPZ, CAPZ's metrics will lose that data. However, that lost data may no longer be meaningful to CAPZ since Azure API calls will be happening out-of-band with CAPZ reconciliations. ASO does measure how many Azure API calls are made and how long each takes and exposes those in the Prometheus format, along with other metrics as documented at https://azure.github.io/azure-service-operator/introduction/metrics/. However, users relying on the current structure of metrics will likely experience breaking changes.
+
+Dependency on ASO may also negatively impact the time between new features being available in Azure and CAPZ being able to take advantage of them. Although ASO can publish updated CRDs with new Azure API versions independently from the SDK, a few factors contribute to the likelihood of ASO lagging a bit behind the SDK:
+- The Azure SDK's code generation pipeline is currently more mature than ASO's.
+- The SDK ships one package at a time whereas ASO ships all resources together, which could result in scenarios where some resources may have updated API versions but are waiting on other unrelated resources to be updated so ASO can batch several updates into a single release.
+- ASO provides some features not included in the SDK, like populating ConfigMaps and Secrets with credentials and other outputs from provisioning, and implementing those features is not automatic.
+- Resource types generated for the SDK occasionally cannot be represented as Kubernetes types for ASO, like resources that are potentially infinitely recursive.
```

Comment:
Are these problematic resource types just wrapper or convenience classes? That is, they don't represent an Azure resource directly.

Do you have an example? I'm just curious.

Reply:
> Are these problematic resource types just wrapper or convenience classes? That is, they don't represent an Azure resource directly.

The examples I am aware of are pretty rare. One is [management groups](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v68.0.0+incompatible/services/resources/mgmt/2020-05-01/managementgroups#CreateManagementGroupChildInfo)

Until/unless CRDs support ref's, it's difficult/impossible to model such types as Kubernetes CRDs. They are quite uncommon (I'm aware of only a few we're found so far), and there are some workarounds.